flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/StreamTaskTest.java

```diff
@@ -1619,6 +1617,50 @@ public void testTaskAvoidHangingAfterSnapshotStateThrownException() throws Excep
         }
     }
 
+    @Test
+    public void testCleanUpResourcesWhenFailingDuringInit() {
+        // given: Configured SourceStreamTask with source which fails during initialization.
+        StreamTaskMailboxTestHarnessBuilder<Integer> builder =
+                new StreamTaskMailboxTestHarnessBuilder<>(
+                                OneInputStreamTask::new, BasicTypeInfo.INT_TYPE_INFO)
+                        .addInput(BasicTypeInfo.INT_TYPE_INFO);
+        try {
+            // when: The task initializing(restoring).
+            builder.setupOutputForSingletonOperatorChain(new InitFailOperator<>()).build();
+            fail("The task should fail during the restore");
+        } catch (Exception ignore) {
+            // ignore.
+        }
+
+        // then: The task should clean up all resources even when it failed on init.
+        assertTrue(InitFailOperator.wasClosed);
+    }
+
+    @Test
+    public void testRethrowExceptionFromRestoreInsideOfInvoke() {
+        // given: Configured SourceStreamTask with source which fails during initialization.
+        StreamTaskMailboxTestHarnessBuilder<Integer> builder =
+                new StreamTaskMailboxTestHarnessBuilder<>(
+                                OneInputStreamTask::new, BasicTypeInfo.INT_TYPE_INFO)
+                        .addInput(BasicTypeInfo.INT_TYPE_INFO);
+        try {
+            // when: The task invocation without preceded restoring.
+            StreamTaskMailboxTestHarness<Integer> harness =
+                    builder.setupOutputForSingletonOperatorChain(new InitFailOperator<>())
+                            .buildUnrestored();
+
+            harness.streamTask.invoke();
+
+            fail("The task should fail during the restore");
+        } catch (Exception ex) {
+            // then: The task should rethrow exception from initialization.
+            assertThat(ex.getMessage(), is(INIT_FAILED_MESSAGE));
```

Comment:
nit/question: I guess here you are freezing a contract that this exception is never wrapped, right? Is this intentional? If not our usual pattern is:
```
if (!ExceptionUtils.findThrowable(e, ExpectedTestException.class).isPresent()) {
    throw e;
}
```